docs/src/keymap-upgrade/parser.ts

```diff
@@ -0,0 +1,95 @@
+import Parser from "web-tree-sitter";
+
+const TREE_SITTER_WASM_URL = new URL(
+  "/node_modules/web-tree-sitter/tree-sitter.wasm",
+  import.meta.url
+);
+
+export let Devicetree: Parser.Language;
+
+export async function initParser() {
+  await Parser.init({
+    locateFile: (path: string, prefix: string) => {
+      // When locating tree-sitter.wasm, use a path that Webpack can map to the correct URL.
+      if (path == "tree-sitter.wasm") {
+        return TREE_SITTER_WASM_URL.href;
+      }
+      return prefix + path;
+    },
+  });
+  Devicetree = await Parser.Language.load("/tree-sitter-devicetree.wasm");
+}
+
+export function createParser() {
+  if (!Devicetree) {
+    throw new Error("Parser not loaded. Call initParser() first.");
+  }
+
+  const parser = new Parser();
+  parser.setLanguage(Devicetree);
+  return parser;
+}
+
+/**
+ * Returns the node for the named capture.
+ */
+export function findCapture(name: string, captures: Parser.QueryCapture[]) {
+  for (const c of captures) {
+    if (c.name === name) {
+      return c.node;
+    }
+  }
+
+  return null;
+}
+
+/**
+ * Returns whether the node for the named capture exists and has the given text.
+ */
+export function captureHasText(
+  name: string,
+  captures: Parser.QueryCapture[],
+  text: string
+) {
+  const node = findCapture(name, captures);
+  return node?.text === text;
+}
+
+/**
+ * Get a list of SyntaxNodes representing a devicetree node with the given path.
+ * (The same node may be listed multiple times within a file.)
+ *
+ * @param path Absolute path to the node (must start with "/")
+ */
+export function findDevicetreeNode(
+  tree: Parser.Tree,
+  path: string
+): Parser.SyntaxNode[] {
+  if (!path.startsWith("/")) {
+    throw new Error("Path must be absolute");
+  }
+
+  const names = ["/", ...path.substring(1).split("/")];
+
+  const query = Devicetree.query(getNodeQuery(names));
+  const matches = query.matches(tree.rootNode);
+
+  return matches.map((match) => findCapture("node", match.captures)!);
+}
+
+function getNodeQuery(names: string[], depth = 0): string {
+  if (names.length === 0) {
+    throw new TypeError();
+  }
+
+  const [name, ...rest] = names;
+
+  const nameCapture = `@_name_${depth}`;
+  const nameQuery = `name: (identifier) ${nameCapture} (#eq? ${nameCapture} "${name}")`;
```

Comment:
How does tree-sitter distinguish node label from name? This doesn't seem to be recognized and replaced:
```dts
mt: behavior_mod_tap
```
while dropping the `mt: ` works fine.

Reply:
Not in a very useful way... It wraps the labeled thing in a new syntax node. That's obnoxious for queries, so I just changed it to add a label field to the labeled thing instead (that change isn't pulled in to ZMK yet), but we'd still run into the same problem in cases like code wrapped by `#if ... #endif`, so I reworked this function to ignore all intermediate syntax nodes that aren't devicetree nodes.

Reply:
Updated tree-sitter-devicetree to 0.9.0. No changes to the behavior here, but it should handle keymaps with `#if` and friends more correctly now.